### ISSUE TITLE
New rule: disallow percent size values that are not `0%` or `100%`.

### ISSUE THREAD

**Issue by giladgd** [stylelint/stylelint#6803]
### What is the problem you're trying to solve?

I would like to have a linting error for percent size values in CSS that are not `0%` or `100%`.
Almost all of the usages I've seen so far of percent size values that are not `0%` or `100%` could have been replaced with a more responsive alternative approach, such as using flexbox and `min`, and `max` size values.
I would like to have a Stylelint rule to discourage using such values in the first place.

Examples of popular bad usages I seen so far:
```scss
.something {
    // using flexbox is better for most cases,
    // as it considers the content of the element and neighboring elements when calculating the size of it.
    // It's possible to also configure flexbox to not consider the content of the element,
    // but when you do need such behavior, you have to deliberately configure it this way
    width: 33%;

    // Centering an element using flexbox is better
    top: 50%;

    // 99% of the usages of this that I've seen so far tried to make an element look circular,
    // but using `999px` is more responsive when the content prevents the element from being circular
    border-radius: 50%;
}
```

### What solution would you like to see?

I would like to have a new rule for this

**Comment by ybiquitous** [comment]
@giladgd Thanks for the proposal with a PoC.

I understand your motivation, but I'm wondering if this rule fits [our following criteria](https://stylelint.io/developer-guide/rules#define-the-rule):

> generally useful; not tied to idiosyncratic patterns

Additionally, the [`declaration-property-value-disallowed-list`](https://stylelint.io/user-guide/rules/declaration-property-value-disallowed-list) rule could likely meet this requirement. For example, see [this demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdilB3AllAC4AW8KAzBQKQDcWOREADuQKwAMdDKARhACdYAgLQCkUfAFcAzuy71MAXywgANCABm+ADYwAckgC2cRDAAex5noB0YGTPXgImbQHMEIDNhToQAqT0ZP3JvHBw-WDAdJHEifBcRZgEWGAEiAE8RADckHSkYEUkZPJ0IXBgoER18GSIQtB5w3xAAegA9AAoCYhIAHyZmPv4hNLEJaRkASgASVoaAbT9WhY4RAE4AXQBqanmQTaaUFR8TpSdIV3w3ADFBIyQiTwArGRcnWGZHRDCWuoy9DVMPUEC0YkQYHU-GoeH5-oD8MCRJd3CI6khMFBYlAGn5wZCQVhzkogA):

```json
{
  "rules": {
    "declaration-property-value-disallowed-list": {
      "/^(width|top|border-radius)$/": ["/[0-9]+%/"]
    }
  }
}
```

That said, I'd like to hear other people's thoughts.

**Comment by giladgd** [comment]
@ybiquitous Thank you for the suggestion. It covers most of the problematic cases I could think of.

I think it may still be a good idea to have a rule for this since people may want to discourage improper usage of percent size values, and having a rule for this will make it more accessible to configure correctly.

---

For the sake of someone who finds this issue, I think this rule will cover most of the problematic use cases I could think of:
```json
{
  "rules": {
    "declaration-property-value-disallowed-list": {
      "/^(background-position|background-size|border-image-width|border-radius|border-top-left-radius|border-top-right-radius|border-bottom-right-radius|border-bottom-left-radius|bottom|column-gap|column-width|flex-basis|font-size|gap|grid-column-gap|grid-row-gap|grid-template-columns|grid-template-rows|height|inset|inset-block|inset-block-start|inset-block-end|inset-inline|inset-inline-start|inset-inline-end|left|line-height|margin|margin-top|margin-right|margin-bottom|margin-left|margin-block|margin-block-start|margin-block-end|margin-inline|margin-inline-start|margin-inline-end|max-height|max-width|min-height|min-width|object-position|padding|padding-top|padding-right|padding-bottom|padding-left|padding-block|padding-block-start|padding-block-end|padding-inline|padding-inline-start|padding-inline-end|right|row-gap|top|transform|vertical-align|width|word-spacing)$/": ["/[\\d.]+%/"]
    }
  }
}
```

> This rule targets only properties that support percent size values.
> It happens that this list doesn't include properties that also accept color as a value, thus this rule will not prevent using a percent value for a color alpha (like `rgba(0 0 0 16%)`)

**Comment by ybiquitous** [comment]
@giladgd Thanks for the helpful example with the demo.

You can also [write a plugin](https://stylelint.io/developer-guide/plugins) and add it to [Awesome Stylelint](https://github.com/stylelint/awesome-stylelint#readme) until the community responds.

**Comment by jeddy3** [comment]
Closing as it's a great candidate for a plugin as it encourages a particular practice.
